torch/autograd/_functions/tensor.py

```diff
@@ -18,9 +18,8 @@ def forward(self, i):
         return result
 
     def backward(self, grad_output):
-        # TODO: this won't have to be zeroed
         grad_input = grad_output.new(self.input_size).zero_()
-        grad_input.index(self.index).copy_(grad_output)
+        grad_input._set_index(self.index, grad_output)
```

Comment:
The long tensor bit needs  a test.
Also, the LongTensor-in-a-tuple case if not allowed has to be asserted in forward